### ISSUE TITLE
Unexpected Anonymous user behavior

### ISSUE THREAD

**Issue by janash** [bprinty/Flask-Authorize#44]
The behavior for anonymous users (`user = None`) seems inconsistent/incorrect to me. I am using the package [flask-jwt-extended](https://flask-jwt-extended.readthedocs.io/en/stable/) for handling user authentication, so when a user is not logged in, `None` is returned for the current user. This is in contrast to `flask-login` where you get an Anonymous User object. I recently switched between the two, and saw no problem when using `flask-login`.

In an app I am creating, users can be logged in or not logged in (public user). Some entries in our DB are public and should be viewable by everyone - all logged in users, and users who are not logged in (public). To me, this is the equivalent of the permission for `other` being set to `read` along with `AUTHORIZE_ALLOW_ANONYMOUS_ACTIONS` being set to `True`.  

I would expect that users who are not logged in experience the same permissions for an item that users who are not the owner or in the group would get (ie `other` permissions). However, if I allow anonymous actions, non-logged in users are able to see *all* entries, even those where `other` is set to no permissions (`other=['']`), which they should not be able to read. I haven't tested, but from looking at the code, `AUTHORIZE_ALLOW_ANONYMOUS_ACTIONS` will allow all actions for users who are not logged in to do any action (delete, create, update, etc).

Relevant code portion from `flask_authorize/plugin.py`, lines `336-338`

~~~
# don't allow anything for anonymous users
        if user is None:
            return current_app.config['AUTHORIZE_ALLOW_ANONYMOUS_ACTIONS']
~~~

In the permission check, if 'authorize_allow_anonymous_actions' is True, using `authorize.action`  will always return True (which seems like unwanted behavior to me). This is also inconsistent with what will come out of the DB if you have `user = None` and do a sqlalchemy query to your db, for example for entries which you are allowed to read. As in, I can do `authorize.read(article)` and get True on an sqlalchemy article object, but do `Article.query.filter(Article.authorized('read'))` and not get the article.

I patched this on my own package to get the desired behavior (https://github.com/janash/seamm_dashboard/blob/jwt-authorize/app/authorize_patch.py). I've tried applying this in the package as well, but it causes many tests to fail. Wondering if you see this behavior as intended. 

If this is unclear, let me know if you need me to come up with a minimal example.

Thanks!

**Comment by bprinty** [comment]
Hi @janash! Thanks for the ticket and thorough explanation.

You make a good point - the logic for handling anonymous users does seem inconsistent with what you'd actually want for content restrictions (defaulting to `other` permissions instead of wide-open permissions). I think your patch resolves the problem appropriately, but I'm going to take some time to go through the failing test cases to ensure they're all logically consistent and expected.

**Comment by bprinty** [comment]
After I do that, I'll create a new release and close this ticket with the new version number. Thanks again!

**Comment by janash** [comment]
Thanks!

**Comment by bprinty** [comment]
Sorry this took so long - it's been a busy couple of weeks and I haven't had time to get around to it until now. I just incorporated your fix and updated the tests accordingly.

**Comment by bprinty** [comment]
I also just released a `0.2.2` version update to PyPi. Thanks for the help!

**Comment by janash** [comment]
👍🏻  👍🏻  👍🏻
